pkg/labels: print all leaf CIDRs, not just the last one.

squeed opened the pull request:
When printing Labels (e.g. by `cilium identity list`), CIDR labels are de-expanded to only show the relevant ones. For example, the CIDR label `1.1.1.0/24` is under-the-hood expanded to `0.0.0.0/0, 0.0.0.0/1, ..., 1.1.0.0/23, 1.1.1.0/24`. Then, when the identity is printed for human consumption, the CIDR labels are de-expanded. However, this code did not anticipate a single identity having multiple "leaf" CIDR labels.

So, change the printing logic to output all interesting CIDRs. This adds a O(N^2) check, but given that N is never over ~100, it shouldn't be an issue.

Note that this output is only for human consumption; it is not persisted or used as an API.

joestringer commented on pkg/labels/labels.go:
```diff
@@ -143,26 +143,40 @@ type Labels map[string]Label
-		res = append(res, cidr)
+
+	// Remove all prefixes that contain a smaller prefix
+	// Yes, this is O(n^2) but N is ~50
```
For one identity. I guess on the commandline we could be printing 1000s of identities? That said as long as it's ideally <100ms even with a large identity list then it's probably in the right order of magnitude for proper interaction with the user. Note this will be hit during sysdump though.

(Would be nice to have a benchmark to evaluate this too..)

squeed replied:
Good idea. This is actually relevant to more than just `ciliumctl`, as it is also used for generating log messages.

I'll do a quick benchmark.

squeed replied:
Wrote a benchmark; decided to make it faster. Good idea.

squeed committed:
pkg/labels: move pkg/labels/cidr up a level

This package implemented a single function and caused import loops. So,
just move it up a level to get rid of the loop.

Signed-off-by: Casey Callendrello <cdc@isovalent.com>

squeed committed:
pkg/labels: print all leaf CIDRs, not just the last one.

When printing Labels (e.g. by `cilium identity list`), CIDR labels are
de-expanded to only show the relevant ones. For example, the CIDR label
1.1.1.0/24 is under-the-hood expanded to 0.0.0.0/0, 0.0.0.0/1, ...,
1.1.1.0/23, 1.1.1.0/24. Then, when the identity is printed for human
consumption, the CIDR labels are de-expanded. However, this code did not
anticipate a single identity having multiple "leaf" CIDR labels.

So, change the printing logic to output all interesting CIDRs. This adds
a small utility datatype to efficiently prune a tree of cidrs to the
leaves. It winds up being about 30% faster than the original, incorrect
version.

Signed-off-by: Casey Callendrello <cdc@isovalent.com>

joestringer reviewed (approved):
Needs a rebase but otherwise LGTM. Nice one :rocket:
